api/utils/keys/yubikey.go

```diff
@@ -0,0 +1,384 @@
+// connections, so we use a retry loop to give the PIV module time to close prior connections.
+func (y *yubiKey) open() (yk *piv.YubiKey, err error) {
+	isRetryError := func(err error) bool {
+		retryError := "connecting to smart card: the smart card cannot be accessed because of other connections outstanding"
+		return strings.Contains(err.Error(), retryError)
+	}
+
+	var maxRetries int = 100
+	for i := 0; i < maxRetries; i++ {
+		yk, err = piv.Open(y.card)
+		if err == nil {
+			return yk, nil
+		}
+
+		if !isRetryError(err) {
+			return nil, trace.Wrap(err)
+		}
+
+		time.Sleep(time.Millisecond * 100)
+	}
+
+	return nil, trace.Wrap(err)
+}
+
+// findYubiKey finds a yubiKey PIV card by serial number. If no serial
+// number is provided, the first yubiKey found will be returned.
+func findYubiKey(serialNumber uint32) (*yubiKey, error) {
+	yubiKeyCards, err := findYubiKeyCards()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	for _, card := range yubiKeyCards {
+		y, err := newYubiKey(card)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		if serialNumber == 0 || y.serialNumber == serialNumber {
+			return y, nil
+		}
+	}
+
+	return nil, trace.NotFound("no yubiKey device found with serial number %q", serialNumber)
+}
+
+// findYubiKeyCards returns a list of connected yubiKey PIV card names.
+func findYubiKeyCards() ([]string, error) {
+	cards, err := piv.Cards()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	var yubiKeyCards []string
+	for _, card := range cards {
+		if strings.Contains(strings.ToLower(card), PIVCardTypeYubiKey) {
+			yubiKeyCards = append(yubiKeyCards, card)
+		}
+	}
+
+	if len(yubiKeyCards) == 0 {
+		return nil, trace.NotFound("no yubiKey devices found")
+	}
+
+	return yubiKeyCards, nil
+}
+
+func parsePIVSlot(slotKey uint32) (piv.Slot, error) {
+	switch slotKey {
+	case piv.SlotAuthentication.Key:
+		return piv.SlotAuthentication, nil
+	case piv.SlotSignature.Key:
+		return piv.SlotSignature, nil
+	case piv.SlotCardAuthentication.Key:
+		return piv.SlotCardAuthentication, nil
+	case piv.SlotKeyManagement.Key:
+		return piv.SlotKeyManagement, nil
+	default:
+		retiredSlot, ok := piv.RetiredKeyManagementSlot(slotKey)
+		if !ok {
+			return piv.Slot{}, trace.BadParameter("slot %X does not exist", slotKey)
+		}
+		return retiredSlot, nil
+	}
+}
+
+// certOrgName is used to identify Teleport Client self-signed certificates stored in yubiKey PIV slots.
+const certOrgName = "teleport"
+
+func selfSignedTeleportClientCertificate(priv crypto.PrivateKey, pub crypto.PublicKey) (*x509.Certificate, error) {
+	serialNumberLimit := new(big.Int).Lsh(big.NewInt(1), 128)
+	serialNumber, err := rand.Int(rand.Reader, serialNumberLimit) // see crypto/tls/generate_cert.go
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	cert := &x509.Certificate{
+		SerialNumber: serialNumber,
+		PublicKey:    pub,
+		Subject: pkix.Name{
```

Comment:
Do we want to store serial and slot in the subject for later debugging, or is it just not necessary?

Reply:
I don't think so, that data should already be known if you're able to find the certificate in the first place